tools/tpm2_print_quote.c

```diff
@@ -0,0 +1,220 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdio.h>
+#include "log.h"
+#include "tpm2_util.h"
+
+static UINT8 fdread_uint8(FILE* fd, bool* read_error) {
+    UINT8 x;
+    if(fread(&x, sizeof(UINT8), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    return x;
+}
+
+static UINT16 fdread_uint16(FILE* fd, bool* read_error) {
+    UINT16 x;
+    if(fread(&x, sizeof(UINT16), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    x = tpm2_util_ntoh_16(x);
+    return x;
+}
+
+static UINT32 fdread_uint32(FILE* fd, bool* read_error) {
+    UINT32 x;
+    if(fread(&x, sizeof(UINT32), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    x = tpm2_util_ntoh_32(x);
+    return x;
+}
+
+static UINT64 fdread_uint64(FILE* fd, bool* read_error) {
+    UINT64 x;
+    if(fread(&x, sizeof(UINT64), 1, fd) != 1) {
+        x = 0;
+        *read_error = true;
+    }
+    x = tpm2_util_ntoh_64(x);
+    return x;
+}
+
+static bool print_hex(FILE* fd, long unsigned int size, bool* read_error) {
+    while(size-- > 0) {
+        printf("%2.2x", (unsigned int)fdread_uint8(fd, read_error));
+        if(*read_error) {
+            printf("\n");
+            return false;
+        }
+    }
+    printf("\n");
+    return true;
+}
+
+static bool print_tpm2b_hex(FILE* fd, bool* read_error) {
+    const UINT16 size = fdread_uint16(fd, read_error);
+    if(*read_error) {
+        return false;
+    }
+    return print_hex(fd, size, read_error);
+}
+
+static bool print_quote(const char* const filename) {
+    bool result = false;
+    bool read_error = false;
+
+    printf("filename=%s\n", filename);
+    FILE* fd = fopen(filename, "rb");
+    if(!fd) {
+        LOG_ERR("%s: Could not open file", filename);
+        return result;
+    }
+
+    // check magic
+    if(fdread_uint32(fd, &read_error) != TPM_GENERATED_VALUE || read_error) {
+        LOG_ERR("%s: Bad magic", filename);
+        goto end;
+    }
+
+    // check type (must be a quote)
+    if(fdread_uint16(fd, &read_error) != TPM_ST_ATTEST_QUOTE || read_error) {
+        LOG_ERR("%s: Not a quote object", filename);
+        goto end;
+    }
```

Comment:
Considering that we have both of these magic values, and this tool could be extended to dump other things, maybe call it tpm2_print <objects> and design a way to make it easy to identify what object you have and look up a parser for it.

Reply:
I agree. Renamed tool to tpm2_print under tools/aux dir with a slightly different API.